### PR TITLE
Esbuild plugin

### PR DESCRIPTION
**Pull request by Toomean**
Hello there! I've worked on a plugin for esbuild( according to #267 ).
Not sure about the needed requirements and the final point for this

Take a look, please

### CODE REVIEW

File: packages/esbuild/package.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@size-limit/esbuild",
+  "version": "6.0.4",
+  "description": "Esbuild plugin for Size Limit",
```
Comment (ai) [inline packages/esbuild/package.json:4]:
```suggestion
  "description": "esbuild plugin for Size Limit",
```

Let’s use downcased name. esbuild’s author prefer it.

**Comment by ai** [comment]
Looks excellent. We just need also `check.import` support to test libraries with tree-shaking.

Search for `check.import` and `import` in the project and you will find a code in `webpack` plugin (and config validation in `size-limit`) which you will need to back port (copy-pasting is OK, but if it will be possible to move some helpers to `size-limit` and then re-use it, it will better).

File: packages/esbuild/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Size Limit Esbuild Plugin
+
+The plugin for [Size Limit] to track dependencies size by Esbuild.
```
Comment (ai) [inline packages/esbuild/README.md:3]: We need to fix `Esbuild` → `esbuild` across all docs, not only in the place where I showed

File: packages/esbuild/package.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  "engines": {
+    "node": "^12.0.0 || ^14.0.0 || >=16.0.0"
+  },
+  "peerDependencies": {
+    "size-limit": "6.0.4"
+  },
+  "dependencies": {
+    "@statoscope/webpack-plugin": "^5.17.0",
+    "css-loader": "^6.5.0",
+    "css-minimizer-webpack-plugin": "^3.0.2",
```
Comment (ai) [inline packages/esbuild/package.json:25]: Why do we need all these dependencies?

File: packages/esbuild/index.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+let { nanoid } = require('nanoid/non-secure')
+let { tmpdir } = require('os')
+let { join, resolve, parse } = require('path')
+let rm = require('size-limit/rm')
+
+let convertConfig = require('./convert-config')
+let runEsbuild = require('./run-esbuild')
+let getConfig = require('./get-config')
+
+const ESBUILD_EMPTY_PROJECT = 0
```
Comment (ai) [inline packages/esbuild/index.js:12]: We will need tests with empty project to validate these constants.

The project with empty file should have `0` size.

The project with one digit export (and `import` config) should have `1` size.

File: packages/webpack/get-config.js [hunk 5]
```diff
@@ -1,27 +1,14 @@
 let StatoscopeWebpackPlugin = require('@statoscope/webpack-plugin').default
-let { writeFile } = require('fs').promises
+let updateCheckWithImports = require('size-limit/updateCheckWithImports')
```
Comment (ai) [inline packages/webpack/get-config.js:2]: I do not use camel case in other files in the project. Be careful.

File: packages/esbuild/package.json (hunk 3 above)
Comment (Toomean) [reply packages/esbuild/package.json:25]: nope, I'm going to remove them

File: packages/esbuild/README.md (hunk 2 above)
Comment (Toomean) [reply packages/esbuild/README.md]: Changed, I've set `"(by |)esbuild",` for the yaspeller dictionary. How can it be properly handled?

File: packages/esbuild/package.json (hunk 3 above)
Comment (Toomean) [reply packages/esbuild/package.json]: Removed

File: packages/webpack/get-config.js (hunk 5 above)
Comment (Toomean) [reply packages/webpack/get-config.js]: Thanks for the point! I will keep an eye on it

**Comment by ai** [comment]
We still need to fix coverage and fix some strange Node.js 12 issue

File: packages/esbuild/index.js (hunk 4 above)
Comment (Toomean) [reply packages/esbuild/index.js:12]: I'm struggling with this a bit. Could you help me with that? Maybe we have an example of how to validate such kinds of constants

Comment (ai) [reply packages/esbuild/index.js:12]: Here is a test example for `*_EMPTY_PROJECT_IMPORT` and `*_EMPTY_PROJECT_IMPORT_GZIP`
https://github.com/ai/size-limit/blob/main/packages/webpack/test/index.test.js#L241-L268

Here is a test example for `*_EMPTY_PROJECT` and `*_EMPTY_PROJECT_GZIP` (yeap, it is in `size-limit` and not in `webpack` plugin for historical reason, feel free to move).
https://github.com/ai/size-limit/blob/main/packages/size-limit/test/run.test.js#L322-L328

**Comment by ai** [comment]
Test coverage shows that we need to calibrate (test) `ESBUILD_EMPTY_PROJECT_IMPORT` and `ESBUILD_EMPTY_PROJECT_IMPORT_GZIP`.

Also, we still have Node.js 12 problem:

```
> error: Could not resolve "/home/runner/work/size-limit/size-limit/packages/esbuild/test/fixtures/unknown.js"

 > error: Refusing to overwrite input file "../../../../../tmp/size-limit-qzG5j_Arw23-2HlwpuSXJ/index.js" (use "allowOverwrite: true" to allow this)

 > error: Refusing to overwrite input file "../../../../../tmp/size-limit-8Rz_8m1YKc_k-mX_IJzsY/index.js" (use "allowOverwrite: true" to allow this)
```

**Comment by Toomean** [comment]
Can't reproduce errors from actions on my local machine, will investigate it on another os. As I see, this problem appears on every node version

**Comment by ai** [comment]
> Can't reproduce errors from actions on my local machine

It could be related to CI environment also

**Comment by ai** [comment]
I pushed merge commit to run CI

**Comment by ai** [comment]
Greta work. Thanks.
